packages/wpcom-proxy-request/src/index.ts

```diff
@@ -178,12 +202,17 @@ const makeRequest = ( originalParams, fn ) => {
  *
  * If no function is specified as second parameter, a promise is returned.
  *
- * @param {object} originalParams - request parameters
- * @param {Function} [fn] - callback response
- * @returns {window.XMLHttpRequest|Promise} XMLHttpRequest instance or Promise
+ * @param originalParams - request parameters
+ * @param fn - callback response
+ *
+ * @returns XMLHttpRequest instance or Promise
  */
-const request = ( originalParams, fn ) => {
-	// if callback is provided, behave traditionally
+export default function request( originalParams: WpcomRequestParams, fn: Callback ): XMLHttpRequest;
+export default function request< T >( originalParams: WpcomRequestParams ): Promise< T >;
+export default function request< T >(
+	originalParams: WpcomRequestParams,
+	fn?: Callback
+): XMLHttpRequest | Promise< T > {
```

Comment:
Will these overloads not cause an issue when trying to call `request(params).then()`? I'm curious how TS will negotiate the missing fn callback, which overload does it choose in that case?

Reply:
When `request` is called with the `fn` callback arg, it returns either `XmlHttpRequest` or `null`. The latter can be returned by `wpcom-xhr-request` when used on Node.js, where there's no XHR. `request` never returns a `Promise` when the `fn` callback is supplied.

Without the `fn` callback, the return value is always a `Promise`.

The returned `XmlHttpRequest` object can be used to subscribe to progress event when uploading or downloading something. There's no other API for that in the `wpcom-*-request` libs.

It would be nice to redesign this API in a major release. Mainly to get rid of exposing the `XmlHttpRequest` object. It's not supported in Node.js, and in `wpcom-proxy-request` the returned instance is somewhat synthetic, supporting event listeners but not much else.

And in the variant that returns a `Promise`, progress event and also aborting requests is not supported at all. `XmlHttpRequest` can do both.

Last note: the type of the `request` function should be defined in the `wpcom` library rather than in `wpcom-proxy-request`. After all, `wpcom-proxy-request` is just a "low level driver" for the `wpcom` library, and `wpcom-xhr-request` and other drivers are supposed to implement the same interface.